include/dlaf/eigensolver/tridiag_solver/kernels.h

```diff
@@ -15,188 +15,126 @@
+  DLAF_ASSERT_HEAVY(n == rank1_tile.size().rows(), n, rank1_tile.size().rows());
+
+  // Copy the bottom row of the top tile or the top row of the bottom tile
+  SizeType row = (is_top_tile) ? m - 1 : 0;
+
+  // Negate the last row of the top eigenvector subproblem matrix (Q1) if rho < 0
+  //
+  // lapack 3.10.0, dlaed2.f, line 280 and 281
+  int sign = (is_top_tile && rho < 0) ? -1 : 1;
+
+  for (SizeType i = 0; i < n; ++i) {
+    rank1_tile(TileElementIndex(i, 0)) = sign * evecs_tile(TileElementIndex(row, i)) / T(std::sqrt(2));
+  }
+}
 
 #ifdef DLAF_WITH_GPU
 
 template <class T>
 void assembleRank1UpdateVectorTile(bool is_top_tile, T rho,
                                    const matrix::Tile<const T, Device::GPU>& evecs_tile,
                                    const matrix::Tile<T, Device::GPU>& rank1_tile,
-                                   whip::stream_t stream);
-
-#define DLAF_GPU_ASSEMBLE_RANK1_UPDATE_VECTOR_TILE_ETI(kword, Type)                        \
-  kword template void assembleRank1UpdateVectorTile(                                       \
-      bool is_top_tile, Type rho, const matrix::Tile<const Type, Device::GPU>& evecs_tile, \
-      const matrix::Tile<Type, Device::GPU>& rank1_tile, whip::stream_t stream)
-
-DLAF_GPU_ASSEMBLE_RANK1_UPDATE_VECTOR_TILE_ETI(extern, float);
-DLAF_GPU_ASSEMBLE_RANK1_UPDATE_VECTOR_TILE_ETI(extern, double);
+                                   whip::stream_t stream) {
+  const SizeType m = evecs_tile.size().rows();
+  const SizeType n = evecs_tile.size().cols();
+  DLAF_ASSERT_HEAVY(n == rank1_tile.size().rows(), n, rank1_tile.size().rows());
 
+  gpu::assembleRank1UpdateVectorTile(is_top_tile, rho, m, n, evecs_tile.ptr(), evecs_tile.ld(),
+                                     rank1_tile.ptr(), stream);
+}
 #endif
 
 DLAF_MAKE_CALLABLE_OBJECT(assembleRank1UpdateVectorTile);
 
-template <class T, Device D, class RhoSender, class EvecsTileSender, class Rank1TileSender>
-void assembleRank1UpdateVectorTileAsync(bool top_tile, RhoSender&& rho, EvecsTileSender&& evecs,
-                                        Rank1TileSender&& rank1) {
-  namespace di = dlaf::internal;
-  auto sender =
-      di::whenAllLift(top_tile, std::forward<RhoSender>(rho), std::forward<EvecsTileSender>(evecs),
-                      std::forward<Rank1TileSender>(rank1));
-  di::transformDetach(di::Policy<DefaultBackend_v<D>>(), assembleRank1UpdateVectorTile_o,
-                      std::move(sender));
-}
-
 #ifdef DLAF_WITH_GPU
-
-template <class T>
-void givensRotationOnDevice(SizeType len, T* x, T* y, T c, T s, whip::stream_t stream);
-
-#define DLAF_GIVENS_ROT_ETI(kword, Type)                                                     \
-  kword template void givensRotationOnDevice(SizeType len, Type* x, Type* y, Type c, Type s, \
-                                             whip::stream_t stream)
-
-DLAF_GIVENS_ROT_ETI(extern, float);
-DLAF_GIVENS_ROT_ETI(extern, double);
-
+using gpu::givensRotationOnDevice;
```

Comment:
Just curious, not necessarily asking for a change: what's the reason for putting this in a nested namespace but then pulling it in here anyway?

Reply:
When compiling without gpu support, the compiler complains that gpu is not a known namespace. E.g. in

https://github.com/eth-cscs/DLA-Future/blob/master/include/dlaf/eigensolver/tridiag_solver/rot.h#L183

Reply:
Ah, I mean: why is it not defined in this namespace to start with instead of pulling it in with `using ...`?

Reply:
That function is a kernel launcher so it belongs to the gpu namespace.
However as it doesn't need an extra wrapper to call it (e.g. no Tile argument) and it cannot be used with the gpu namespace, `using` is the nicer solution.

Reply:
All right, thanks for the explanation. Makes sense.